sip/sip-004-materialized-view.md

```diff
@@ -0,0 +1,822 @@
+_O(B)_ in the worst case (where _B_ is the number of blocks), the number of rows
+a peer will visit when resolving a back-pointer can be be at most _O(log B)_ -- i.e.
+this would only happen if the blockchain's forks were organized into a
+perfectly-balanced binary tree. 
+
+In practice, there will be one _long_ fork row that encodes the canonical
+history, as well as number of short fork rows that encode short-lived forks
+(which can arise naturally from burn chain reorganizations).  This means
+resolving back-pointers while working on the longest fork -- the fork where a
+miner's block rewards are most likely to be realized -- will be _O(1)_ in
+expectation.  To help achieve this, the ancestor table would be implemented as
+a hash table in order to ensure that finding the ancestor block also runs in
+_O(1)_ time.
+
+### Merklized Skip-list
+
+The third principal data structure in a MARF is a Merklized skip-list encoded
+from the block header hashes and ART root hashes in each block.  The hash of the
+root node in the ART for block _N_ is derived not only from the hash of the
+root's children, but also from the hashes of the block headers from blocks
+`N - 1`, `N - 2`, `N - 4`, `N - 8`, `N - 16`, and so on.  This constitutes
+a _Merklized skip-list_ over the sequence of ARTs.
+
+The reason for encoding the root node's hash this way is to make it possible for
+peers to create a cryptographic proof that a particular key maps to a particular
+value when the value lives in a prior block, and can only be accessed by
+following one or more back-pointers.  In addition, the Merkle skip-list affords
+a client _two_ ways to verify key-value pairs:  the client only needs either (1)
+a known-good root hash, or (2) the sequence of block headers for the Stacks
+chain and its underlying burn chain.  Having (2) allows the client to determine
+(1), but calculating (2) is expensive for a client doing a small number of
+queries.  For this reason, both options are supported.
+
+### MARF Merkle Proofs
+
+A Merkle proof for a MARF is constructed using a combination of two types of
+sub-proofs:  _segment proofs_, and _shunt proofs_.  A _segment proof_ is a proof
+that a node belongs to a particular Merklized ART.  It is simply a Merkle tree
+proof.  A _shunt proof_ is a proof that the ART for block _N_ is exactly _K_
+blocks away from the ART at block _N - K_.  It is generated as a Merkle proof
+from the Merkle skip-list.
+
+Calculating a MARF Merkle proof is done by first calculating a segment proof for a
+sequence of path prefixes, such that all the nodes in a single prefix are in the
+same ART.  To do so, the node walks from the current block's ART's root node
+down to the leaf in question, and each time it encounters a back-pointer, it
+generates a segment proof from the _currently-visited_ ART to the intermediate
+node whose child is the back-pointer to follow.  If a path contains _i_
```

Comment:
Okay, in my mind, there's two things that could happen in the event of _following_ a back-point (which I think is the only interaction with backpointers that matters while generating a segment proof -- if it's just a sibling node to the path being proved, you just simply include it's pruned hash). It's not clear which of those is being described:

1. When you follow a back-pointer, you treat the back-pointer like any other node in the Merkle path -- basically you hash the contents of the linked node and continue processing.
2. When you follow a back-pointer to block _m_, you create a new Merkle path from the root of _m_ to the intermediate node you followed.

Reading back over the description of using the _block hash_ for back pointer's hashes, I think the answer is (2). Is this is also why shunt proofs are required as described: you need to prove the inclusion (and block height) of each block used in the segment proofs?

As a further test of my understanding, the reason for doing this is to avoid copying the node hash in addition to the blockhash (which is necessary for lookups in the fork table in any case)?

Reply:
The answer is (2).  In verifying a back-pointer from block _B_ and node _N_ to block _B - K_ to node _N'_, it is sufficient to:

* prove that block _B - K_'s root hash is, in fact, _K_ blocks prior to _B_ (using a shunt proof)
* prove that node _N'_ is a child of the root of _B - K_, and is located in a path that is prefixed with the path to node _N_ from _B_ (using a segment proof).

> As a further test of my understanding, the reason for doing this is to avoid copying the node hash in addition to the blockhash (which is necessary for lookups in the fork table in any case)?

That is correct.  The fork table can fit entirely in RAM for the foreseeable future, so inserting the hash of block _B - K_ as the hash of _N'_ when calculating the root hash of _B_ does not carry any I/O penalty.  The naive way to go about this is to simply load the hash of _N'_'s children in _B - K_, but this would take no fewer than _C+1_ disk seeks for _C_ children of _N'_ (at least 1 seek is required to load _N'_'s data).